### PR TITLE
ref(compactSelect): Add optional toggle all buttons

### PR DESCRIPTION
**Pull request by vuluongj20**
Add a button to select/unselect all options within a `CompactSelect` section:
Screenshot 2023-02-13 at 3 07 08 PM

The implementation is a bit complicated than expected because of restrictions on keyboard interactions within list boxes. We need two sets of toggle buttons: one _visible_, _clickable_ but _not keyboard-focusable_ set nested within the list box (this is what you see in the screenshot), and another _hidden_ but _keyboard-focusable_ set outside the list box. The latter is necessary for accessibility reasons and appears in the accessibility tree:
Screenshot 2023-02-13 at 3 12 04 PM

### CODE REVIEW

File: static/app/components/compactSelect/control.tsx [hunk 1]
```diff
@@ -399,8 +399,7 @@ const MenuHeader = styled('div')<{size: FormSize}>`
   display: flex;
   align-items: center;
   justify-content: space-between;
-  padding: ${p => headerVerticalPadding[p.size]} ${space(1)}
-    ${p => headerVerticalPadding[p.size]} ${space(1.5)};
+  padding: ${p => headerVerticalPadding[p.size]} ${space(1.5)};
```
Comment (vuluongj20) [inline static/app/components/compactSelect/control.tsx:402]: Update padding so the vertical keylines match with that of section headings.

File: docs-ui/stories/components/compactSelect.stories.js [hunk 2]
```diff
@@ -92,3 +83,35 @@ _CompactSelect.argTypes = {
...
+        showToggleAllButton: true,
+        options: [
+          {value: 'choice_three', label: 'Choice Three'},
+          {value: 'choice_four', label: 'Choice Four'},
+        ],
+      },
+    ]}
+    {...props}
+  />
+);
```
Comment (vuluongj20) [inline docs-ui/stories/components/compactSelect.stories.js:110]: Add a "Grouped" story to better document the differences between `CompactSelect`s with grouped options vs those with a linear array of options.

File: static/app/components/compactSelect/listBox.tsx [hunk 3]
```diff
@@ -288,12 +311,98 @@ function ListBox<Value extends React.Key>({
...
+      const visibleCounterpart = listBoxRef.current?.querySelector(
+        `li[data-key="${item.key}"] button[aria-hidden]`
+      );
+
+      if (!visibleCounterpart) {
+        return;
+      }
+      visibleCounterpart.classList.remove('focus-visible');
+    },
+  });
```
Comment (vuluongj20) [inline static/app/components/compactSelect/listBox.tsx:377]: The result:
https://user-images.githubusercontent.com/44172267/218596994-283446e5-10f1-44fb-a78d-2dd49ee7c721.mov

**Review by malwilley — COMMENTED** [review]
Great work navigating all the accessibility challenges here!

Code looks good to me, but do we want some tests for this?

**Comment by vuluongj20** [comment]
Marked as draft as I want to merge https://github.com/getsentry/sentry/pull/45003 in first (I'll need to update the code in this PR following the changes made in https://github.com/getsentry/sentry/pull/45003).

**Comment by vuluongj20** [comment]
> Great work navigating all the accessibility challenges here!
> 
> Code looks good to me, but do we want some tests for this?

@malwilley Oh true! Added some tests in https://github.com/getsentry/sentry/pull/44547/commits/406c59fc00cdfd7ad937c8c9ce244de9c70c9928

File: static/app/components/compactSelect/list.tsx [hunk 4]
```diff
@@ -281,76 +282,47 @@ function List<Value extends React.Key>({
...
   );
 }
 
 export {List};
-
-/**
- * Recursively finds the selected option(s) from an options array. Useful for
- * non-flat arrays that contain sections (groups of options).
- */
-function getSelectedOptions<Value extends React.Key>(
```
Comment (vuluongj20) [inline static/app/components/compactSelect/list.tsx:313]: Moved to `./utils`

**Review by malwilley — APPROVED** [review]
Looks good but lint seems to be failing?

**Pull request merged**